trieTools/trieToolsCommon/addressTokens.go

```diff
@@ -0,0 +1,103 @@
+package trieToolsCommon
+
+// This is not concurrent safe
+type addressTokensMap struct {
+	internalMap map[string]map[string]struct{}
+}
+
+func NewAddressTokensMap() AddressTokensMap {
+	return &addressTokensMap{
+		internalMap: make(map[string]map[string]struct{}),
+	}
+}
+
+// Add will add all provided tokens to the corresponding address
+func (atm *addressTokensMap) Add(address string, tokens map[string]struct{}) {
+	_, addressExists := atm.internalMap[address]
+	if !addressExists {
+		atm.internalMap[address] = make(map[string]struct{})
+	}
+
+	atm.addTokens(address, tokens)
+}
+
+func (atm *addressTokensMap) addTokens(address string, tokens map[string]struct{}) {
+	tokensCopy := copyTokens(tokens)
```

Comment:
It is mandatory to copy here, since the map is passed as reference and later modified (merges, deletion, addition). This fixes a bug I've been hunting for few hours

Reply:
right. The copy is done in the 3 lines below. I was mentioning the first copy call on L25. That is useless: you copy the provided map to create a full copy of the map called tokensCopy that you iterate again in the for...range to add the elements to the inner map
```GO
func (atm *addressTokensMap) addTokens(address string, tokens map[string]struct{}) {
    for token := range tokens {
        atm.internalMap[address][token] = struct{}{}
    }
}
```

Reply:
Right, I've totally misunderstood it. Fixed as proposed